File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys; input = sys.stdin.readline
from collections import deque
N = int(input())
graph = [list(map(int,input().rstrip())) for _ in range(N)]

def dfs(x, y):
    global num

    if x<0 or x >= N or y < 0 or y >= N:
        return False

    if graph[x][y] == 1:
        graph[x][y] = 0
        num+=1
        dfs(x + 1, y)
        dfs(x - 1, y)
        dfs(x, y + 1)
        dfs(x, y - 1)

        return True
    return False

result = 0
ans = []
num = 0
for i in range(N):
    for j in range(N):
        if dfs(i,j) == True:
            result += 1
            ans.append(num)
            num = 0
print(result)
ans.sort()
print(*ans,sep = '\n')

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;

public class Main {
    static int n;
    static int[][] arr;
    static boolean[][] visit;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        visit = new boolean[n][n];

        for (int i = 0; i < n; i++) {
            String s = br.readLine();
            for (int j = 0; j < n; j++) {
                arr[i][j] = s.charAt(j) - '0';
            }
        }

        int ans = 0;
        ArrayList<Integer> list = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(arr[i][j] == 1 && !visit[i][j]){
                    int cnt = dfs(i, j);
                    list.add(cnt);
                    ans++;
                }
            }
        }
        Collections.sort(list);
        sb.append(ans + "\n");

        for (int i = 0; i < list.size(); i++) {
            sb.append(list.get(i) + "\n");
        }
        System.out.println(sb);
    }

    private static int dfs(int x, int y) {
        if(x < 0 || x >= n || y < 0 || y >= n || arr[x][y] == 0 || visit[x][y]){
            return 0;
        }
        visit[x][y] = true;
        int cnt = 1;

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            cnt += dfs(nx, ny);
        }
        return cnt;
    }
}
